core/src/main/java/org/apache/calcite/plan/RelOptPlanner.java

```diff
@@ -350,4 +350,49 @@ public CannotPlanException(String message) {
       super(message);
     }
   }
+
+  /**
+   * Converts a relation expression to a given set of traits, if it does not
```

Comment:
I am wondering if it is a good idea to add yet another `convert` method or if it would be better to put this logic into the existing `changeTraits` method inside the planner and change rules to call `changeTraits` directly.

The `VolcanoPlanner#changeTraits` already has some assertions that the traits are different so it may make sense to relax it and bail out instead. 

The `HepPlanner#changeTraits` does not care much about traits so we should be good.

Reply:
I'm a bit concerned of potential side-effects on the existing callers of planner `changeTraits`, which currently don't go through this logic, if they might get impacted if they start going through it...

Reply:
I think the risk is small given that we have the assertions about traits in the Volcano planner. I feel that having less APIs for extension and maintenance is positive for the future of the project but maybe other have a different opinion; feel free to open the discussion to the dev list if you like.

If we decide to go for new methods then I guess I like more your suggestion about adding a new `RelOptRule#convert` with an additional planner parameter.

Reply:
@zabetak I have applied your idea on the second commit:
- Move convert logic out of `RelOptPlanner` back to `RelOptRule`
- Inside rule `onMatch`, call directly planner `changeTraits` instead of convert
- Relax assertions inside `VolcanoPlanner#changeTraits`

wdyt?

Reply:
Thanks for the feedback @zabetak !
I'm looking again into this, and currently there are a lot of ConverterRules that still call `RelOptRule#convert`, so in order to deprecate it, we should refactor those too (not covered by the PR as it is right now). 
In any case, I think we might want to refactor them and avoid calling `RelOptRule#convert` in their case too, because theoretically the same problem could happen on them (maybe it is a bit more far-fetched to see it in a real production env, but I have been able to force it in a UT). I don't have much time atm to re-work this, I'll try to circle back to it in the coming days...

Reply:
@zabetak  I've created a new PR (https://github.com/apache/calcite/pull/3243) with the idea of new RelOptRule#convert with an additional planner parameter, which perhaps is the most pragmatic and less disruptive solution to correct this problem, wdyt?